### DISCUSSION TITLE
Automatically Set Binding in Behaviors

### DISCUSSION THREAD

**Discussion by david-maw** [CommunityToolkit/Maui#2417] [New Feature Discussions]
The new .NET 9 implementation of not setting BindingContext in a behavior is a pain, it leads to XAML like:
``` XAML
BindingContext="{Binding BindingContext, Source={x:Reference page}, x:DataType=ContentPage}"
```

Would it be reasonable to set Biding automatically using something like this (this example is for an Entry, but I *think* it can be generalized:
``` C#
    bool bindingWasSet = false;
    Entry? savedEntry;
    protected override void OnAttachedTo(Entry entry)
    {
        savedEntry = entry;
        if (bindingWasSet = (BindingContext is null))
            SetBinding(BindingContextProperty,
            new Binding
            {
                Path = BindingContextProperty.PropertyName,
                Source = entry,
            });
        ... insert behavior initialization code here...
        base.OnAttachedTo(entry);
    }

    protected override void OnDetachingFrom(Entry entry)
    {
        if (bindingWasSet)
            BindingContext = null;
        ... insert behavior termination code here...
        savedEntry = null;
        base.OnDetachingFrom(entry);
    }
```
This just sets the binding to that of the parent if it is null at attach time and remembers it did it so it can be undone when detached.

It would obscure the fact that BindingContext is not set by default for .NET 9 Behaviors, but I don't really understand the argument for not setting BindingContext, since almost any property could be set in a style. It would be better to do this in the framework, but that's what just changed!

**Comment by pictos** [comment]
@david-maw since Xamarin.Forms Behaviors doesn't set the BindingContext automatically.

You can see that info in [docs](https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/behaviors?view=net-maui-8.0#create-a-net-maui-behavior)

**Comment by david-maw** [comment]
Sorry, I'd assumed the Community Toolkit change was reflecting some change in .NET MAUI, obviously not. That means my suggestion comes down to "please don't introduce this breaking change in the Community Toolkit Behaviors".

Do you (or anyone for that matter) know what it is about Behaviors that makes it problematic for them to inherit BindingContext? I get that they can be specified in Style entries, but so can many other properties, so what's special about Behaviors?

**Reply by pictos** [reply]
I don't know all the underlines about it... What we can ask is for Maui team to upgrade `PlatformBehavior` and allow it to work with Bindings... The way we implemented it we found a lot of bugs in virtualized controls, such ListView and CollectionView. Now we defer the responsibility to set the BindingContext to devs

**Reply by jnromeroYT** [reply]
awful

**Reply by jnromeroYT** [reply]
awful change

**Reply by jnromeroYT** [reply]
awful change

**Reply by jnromeroYT** [reply]
bad change
